SU2_CFD/include/drivers/CDiscAdjMultizoneDriver.hpp

```diff
@@ -27,10 +27,41 @@
 
 #pragma once
 #include "CMultizoneDriver.hpp"
+#include "../../../Common/include/toolboxes/CQuasiNewtonInvLeastSquares.hpp"
+#include "../../../Common/include/linear_algebra/CPreconditioner.hpp"
+#include "../../../Common/include/linear_algebra/CMatrixVectorProduct.hpp"
+#include "../../../Common/include/linear_algebra/CSysSolve.hpp"
 
 class CDiscAdjMultizoneDriver : public CMultizoneDriver {
 
 protected:
+#ifdef CODI_FORWARD_TYPE
+  using Scalar = su2double;
+#else
+  using Scalar = passivedouble;
+#endif
+
+  class AdjointProduct : public CMatrixVectorProduct<Scalar> {
+  public:
+    CDiscAdjMultizoneDriver* const driver;
+    const unsigned short iZone = 0;
+    mutable unsigned long iInnerIter = 0;
+
+    AdjointProduct(CDiscAdjMultizoneDriver* d, unsigned short i) : driver(d), iZone(i) {}
+
+    inline void operator()(const CSysVector<Scalar> & u, CSysVector<Scalar> & v) const override {
+      driver->SetAllSolutions(iZone, true, u);
+      driver->Iterate(iZone, iInnerIter, true);
+      driver->GetAllSolutions(iZone, true, v);
+      v -= u;
```

Comment:
The "Krylov discrete adjoint" part is for the inner iterations of multizone problems.
It is based on manipulating the fixed point `lambda = Ext + lambda (dx G)` into `lambda (dx G + I) = -Ext`, and calling the LHS a "product" which is passed to GMRES.
This is not a strictly valid thing to do since there is a kind of flexible preconditioner in G, but it works better than the quasi-Newton thing (which is still available).
This manipulation is not possible for single zone because the RHS of the adjoint problem (i.e. the OF gradient) is not separate from the adjoints of the iteration.

Reply:
Please keep the quasi-Newton thing for a while, I'm still building the "real"-Newton thing (through a class inheriting from `CQuasiNewtonInvLeastSquares`; one would have to figure out a common base class). Interestingly the real one requires the computation of Krylov subspaces, too, so there might be new or other overlaps..

Reply:
Of course, I'm a hoarder of features, I keep everything I can.

Reply:
Alright. From what I see, most of the implementation happens within the new integration class with little interference with the solver code. I'd have to read carefully through the first paper but I can also do it post mortem ;-)
I'll create a WIP for the Newton corrector in the next few days so that you can check what I'm up to there. I think it's quite different from what is done here though :-)